### PR TITLE
Prep for Docker log drivers announcement 

### PR DESCRIPTION
**Pull request by tgross**
@misterbisson I'm opening this PR but it's still work-in-progress.
The current master branch prior to merging this was tagged as `preview`.

Changes:
- check package signatures for Logstash and Containerbuddy during build
- discover Consul via Triton CNS when deployed to Triton (but works locally via envvar)
- consolidate the Docker Compose files to take advantage of getting rid of `--link`s in the main file
- split running the stack (which can be done with just `docker-compose up -d`) from testing the stack
- pull "assert the world" code from Triton Touchbase blueprint into the `./test.sh check` function
- lots of README updates

### CODE REVIEW

**Comment by tgross** [comment]
@misterbisson this and https://github.com/tgross/triton-elasticsearch/pull/2 are probably ready for re-review at this point.

File: README.md [hunk 1]
```diff
@@ -1,76 +1,106 @@
...
+This repo is a demonstration of the [Elasticsearch-Logstash-Kibana (ELK)](https://www.elastic.co/products) stack, designed to be self-operating according to the [autopilot pattern](http://autopilotpattern.io/). This application also demonstrates Triton support of Docker log drivers.
+
+The components of this stack are:
+
+- [Elasticsearch](https://www.elastic.co/products/elasticsearch), to store and search our logs.
+- [Logstash](https://www.elastic.co/products/logstash), to ingest logs from our containers and write them to Elasticsearch.
+- [Kibana](https://www.elastic.co/products/kibana), as a web UI to search the logs we store in Elasticsearch.
+- [Consul](https://www.consul.io/), acting a service catalog to support discovery
+- [Containerbuddy](http://containerbuddy.io), to help with service discovery and bootstrap orchestration
+- [Triton](https://www.joyent.com/), Joyent's container-native infrastructure platform, including the new [Docker log drivers](https://www.joyent.com/blog/docker-log-drivers) support.
```
Comment (misterbisson) [inline README.md]: Full stop with just the following, maybe?

> [Triton](https://www.joyent.com/), Joyent's container-native infrastructure platform

File: README.md [hunk 2]
```diff
@@ -1,76 +1,106 @@
...
+- Nginx, acting as a source of logs for testing.
```
Comment (misterbisson) [inline README.md]: Poor Nginx gets no link. sadface.

File: README.md [hunk 3]
```diff
@@ -1,76 +1,106 @@
...
 1. [Get a Joyent account](https://my.joyent.com/landing/signup/) and [add your SSH key](https://docs.joyent.com/public-cloud/getting-started).
-1. Install the [Docker Toolbox](https://docs.docker.com/installation/mac/) (including `docker` and `docker-compose`) on your laptop or other environment, as well as the [Joyent CloudAPI CLI tools](https://apidocs.joyent.com/cloudapi/#getting-started) (including the `smartdc` and `json` tools).
-1. Make sure GNU make is installed.
+1. Install the [Docker Toolbox](https://docs.docker.com/installation/mac/) (including `docker` and `docker-compose`) on your laptop or other environment, as well as the [Joyent Triton CLI](https://www.joyent.com/blog/introducing-the-triton-command-line-tool) (`triton` replaces our old `sdc-*` CLI tools)
+1. [Configure Docker and Docker Compose for use with Joyent](https://docs.joyent.com/public-cloud/api-access/docker):
 
-### Running the test rig
+```bash
+curl -O https://raw.githubusercontent.com/joyent/sdc-docker/master/tools/sdc-docker-setup.sh && chmod +x sdc-docker-setup.sh
+./sdc-docker-setup.sh -k us-east-1.api.joyent.com <ACCOUNT> ~/.ssh/<PRIVATE_KEY_FILE>
```
Comment (misterbisson) [inline README.md:32]: I think you're missing a closing fence right below here.

File: test.sh [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        echo
+        exit 1
+    fi
+}
+
+# get the IP:port of a container via either the local docker-machine or from
+# triton CLI
+getIpPort() {
+    if [ -z "${COMPOSE_FILE}" ]; then
+        local ip=$(triton inst get ${COMPOSE_PROJECT_NAME}_$1_1 | json -a ips.1)
```
Comment (misterbisson) [inline test.sh]: No desire to use the CNS name?

File: README.md [hunk 5]
```diff
@@ -1,76 +1,106 @@
...
+
+- [Elasticsearch](https://www.elastic.co/products/elasticsearch), to store and search our logs.
+- [Logstash](https://www.elastic.co/products/logstash), to ingest logs from our containers and write them to Elasticsearch.
+- [Kibana](https://www.elastic.co/products/kibana), as a web UI to search the logs we store in Elasticsearch.
+- [Consul](https://www.consul.io/), acting a service catalog to support discovery
+- [Containerbuddy](http://containerbuddy.io), to help with service discovery and bootstrap orchestration
+- [Triton](https://www.joyent.com/), Joyent's container-native infrastructure platform, including the new [Docker log drivers](https://www.joyent.com/blog/docker-log-drivers) support.
+- Nginx, acting as a source of logs for testing.
 
 ![Diagram of Triton-ELK architecture](./doc/triton-elk.png)
```
Comment (misterbisson) [inline README.md:18]: There's something about this diagram that makes me think that Logstash is a back-end for Nginx. That interpretation isn't easily justified, as the relationship lines clearly do not always indicate "back-end," as the lines to Consul should make clear, it's just that seeing Nginx without a back-end is so unusual.

File: test.sh [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+# ---------------------------------------------------
+# utility functions
+
+# check for prereqs
+check() {
+    command -v docker >/dev/null 2>&1 || {
+        echo
+        tput rev  # reverse
+        tput bold # bold
+        echo 'Docker is required, but does not appear to be installed.'
```
Comment (misterbisson) [inline test.sh:107]: Don't forget to `tput sgr0 # clear` after this.

**Comment by misterbisson** [comment]
Perhaps this needs a mention of the `_env` file and why it's there?

_And..._ perhaps we should give the user a little help in gearing the `_env` file?

``` bash
local account_uuid=$(triton account get | awk -F": " '/id:/{print $2}')
local target_dc=$(triton profile get | awk -F"/" '/url:/{print $3}' | awk -F'.' '{print $1}')

CONSUL=consul.svc.${account_uuid}.{target_dc}.cns.joyent.com
```

File: test.sh (hunk 4 above)
Comment (tgross) [inline test.sh]: I can do that... just need to add CNS labels to Nginx. Probably will change the function name to `getPublicUrl` or something then (I want it to support local development too).

File: test.sh (hunk 6 above)
Comment (tgross) [inline test.sh:107]: Good catch.

File: README.md (hunk 3 above)
Comment (tgross) [inline README.md:32]: :+1: will fix

**Comment by tgross** [comment]
> Perhaps this needs a mention of the _env file and why it's there?
> And... perhaps we should give the user a little help in gearing the _env file?

Good call on both.

File: README.md (hunk 5 above)
Comment (tgross) [inline README.md:18]: Cluttering it with something that we're not supplying here doesn't seem like an awesome idea either though. What if we just mark those arrows with the label "logs"?

Comment (misterbisson) [inline README.md:18]: Labeling the arrows would definitely help, but then I got concerned that we might need to label all the arrows. 

I also wondered if the diagram made Nginx appear to be too much a part of the ELK composition, and maybe the two should be segregated in separate groups: box around ELK components, box around log generating components.

**Comment by misterbisson** [comment]
> gearing

Damn you autocorrect!

**Comment by tgross** [comment]
Updated that diagram so I think this is ready for one more pass.

File: test.sh [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        exit
+    fi
+do
+    echo
+done
+
+# default behavior
+echo "Starting example application"
+echo "project prefix:      $COMPOSE_PROJECT_NAME"
+echo "docker-compose file: $COMPOSE_FILE"
```
Comment (misterbisson) [inline test.sh]: `$COMPOSE_FILE` is empty for the `./test.sh` with no args case, resulting in empty output.

File: test.sh [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+TRITON_USER=
+TRITON_DC=
+TRITON_ACCOUNT=
+
+# ---------------------------------------------------
+# Top-level commmands
+
+run() {
+    check
+    docker-compose up -d \
```
Comment (misterbisson) [inline test.sh]: Consul is missing from this list.

Does it make sense to name each service? Or is a `docker-compose up -d` enough?

Comment (misterbisson) [inline test.sh]: It's likely not necessary (given that the `elasticsearch` node can provide the data service), but on testing it seems `elasticsearch_data` is missing from the list as well.

File: test.sh [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+
+# Check for correct configuration
+check() {
+    command -v docker >/dev/null 2>&1 || {
+        echo
+        tput rev  # reverse
+        tput bold # bold
+        echo 'Docker is required, but does not appear to be installed.'
+        echo 'See https://docs.joyent.com/public-cloud/api-access/docker'
+        tput sgr0 # clear
```
Comment (misterbisson) [inline test.sh:108]: Minor inconsistency: the other error messages just highlight one line, not the second information line

File: test.sh [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    echo "$2"
+    while :
+    do
+        curl --fail -s -o /dev/null "$1" && break
+        sleep 1
+        echo -ne .
+    done
+    echo
+    echo "$3"
+    open "$1"
```
Comment (misterbisson) [inline test.sh]: Do you want to test for `open`, so we don't emit errors here when not on a Mac?

File: test.sh [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+
+# ---------------------------------------------------
+# utility functions
+
+# get the address for a container via either local docker-machine or Triton CLI.
+# requires service name (or name + port + protocol for local use)
+getPublicUrl() {
+    if [ -z "${COMPOSE_FILE}" ]; then
+        local cname=$(printf '%s.svc.%s.%s.triton.zone' $1 ${TRITON_ACCOUNT} ${TRITON_DC})
+        echo "$cname:$port"
```
Comment (misterbisson) [inline test.sh:180]: `$port` is `$2`?

Comment (tgross) [inline test.sh]: This should be: `echo "$cname:$2"`

File: test.sh (hunk 7 above)
Comment (tgross) [inline test.sh]: That's how we've done most of these... if you don't pass COMPOSE_FILE then Compose picks the `docker-compose.yml` by default. I can have it write out the value here instead if you think it'll be less confusing.

Comment (misterbisson) [inline test.sh]: Yeah, I'm feeling we should either emit the file we're using here or conditionally not emit the static text label.

File: test.sh (hunk 8 above)
Comment (tgross) [inline test.sh]: Yah, that was left over from the original version where we had the Nginx stuffed into the same `docker-compose.yml` file. Will fix.

**Comment by tgross** [comment]
I've picked up those comments in https://github.com/tgross/triton-elk/pull/2/commits/bf24c3526bbdcdc3866887a41c0154ce2042be84

File: test.sh [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        echo
+        tput rev  # reverse
+        tput bold # bold
+        echo 'Error! Triton CNS is required and not enabled.'
+        tput sgr0 # clear
+        echo
+        exit 1
+    fi
+
+    echo CONSUL=consul.svc.${TRITON_ACCOUNT}.${TRITON_DC}.triton.zone > _env
```
Comment (misterbisson) [inline test.sh:168]: I'm agreeing that it is very strange this did not work as `...cns.joyent.com`, as the private name should be usable by everything on the same fabric.

File: nginx/containerbuddy.json [hunk 13]
```diff
@@ -1,5 +1,5 @@
 {
-  "consul": "consul:8500",
+  "consul": "{{ .CONSUL }}:8500",
   "onStart": "/opt/containerbuddy/reload.sh",
```
Comment (misterbisson) [inline nginx/containerbuddy.json]: This calls https://github.com/tgross/triton-nginx/blob/master/bin/reload.sh#L19, causing it to try `nginx -s reload` before the Nginx starts, triggering an error:

```
2016/03/17 18:53:53 [error] 11366#11366: open() "/var/run/nginx.pid" failed (2: No such file or directory)
nginx: [error] open() "/var/run/nginx.pid" failed (2: No such file or directory)
```

File: test.sh [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+        # https://github.com/logstash-plugins/logstash-codec-fluent/pull/5
+        # fluentd)
+        #    port=24224 ;;
+        *)
+            echo 'logtype arguments required: gelf or syslog'
+            exit 1;;
+    esac
+
+    echo 'Starting Nginx log source...' && \
+        CONSUL=$(getPrivateIpPort consul 8500 tcp) \
```
Comment (misterbisson) [inline test.sh]: Does it make sense to get `CONSUL` from here, or should we use the value from `_env`? The inconsistency seems unnecessary.

File: nginx/containerbuddy.json (hunk 13 above)
Comment (tgross) [inline nginx/containerbuddy.json]: Noted in https://github.com/tgross/triton-nginx/issues/6 and https://github.com/tgross/workshop-autopilot/issues/2. We'll need to update in the Docker image and here in the Containerbuddy config.

File: test.sh (hunk 14 above)
Comment (tgross) [inline test.sh]: When running locally we need to use links and not CNS, so the `_env` file which gets used by `docker-compose.yml` is populated by CNS while the `local-compose.yml` file just uses links and the static string "consul". If all we wanted to do was stand up some containers we could just do all this logic in two compose files, but we're doing things like opening the web interfaces.

~~You've asked about this a couple different ways now, so let me propose this approach:~~
~~\- we ditch the `test.sh` file entirely~~
~~\- we provide a `docker-compose.yml`, `local-compose.yml`, `test-compose.yml`, and `test-local-compose.yml` that does all the standing up of things.~~
~~\- we provide a `check.sh` file to do the environment checking and open web portals, but it only works on Triton and not locally.~~

Nevermind, we have to inject the Nginx config and Containerbuddy config into the `0x74696d/triton-nginx` container, which we can't do via Compose.

Comment (misterbisson) [inline test.sh]: This is a-ok as is.

File: test.sh (hunk 12 above)
Comment (misterbisson) [inline test.sh:168]: We learned that one of the DCs did not match the expected pattern. That's changed or changing. Should we switch this back?

Comment (tgross) [inline test.sh:168]: Let's avoid changing it until after it's done, so that we don't get folks from inside Joyent filing bugs in the meantime.

**Comment by misterbisson** [comment]
:house_with_garden: :walking:

**Pull request merged**